botfront/imports/ui/components/templates/templates-list/BotResponseEditor.jsx

```diff
@@ -0,0 +1,322 @@
+            },
+            (error) => {
+                callback(error);
+            },
+        );
+    };
+
+    const handleChangeMetadata = (updatedMetadata) => {
+        if (isNew) {
+            setNewBotResponse(
+                { ...(newBotResponse || botResponse), metadata: updatedMetadata });
+            return;
+        }
+        updateResponse({ ...botResponse, metadata: updatedMetadata }, () => {});
+    };
+
+    const handleChangeKey = async () => {
+        if (!responseKey.match(/^utter_/)) {
+            setRenameError('Response names must start with "utter_"');
+            return;
+        }
+        if (isNew) {
+            setNewBotResponse({ ...(newBotResponse || botResponse), key: responseKey });
+            return;
+        }
+        updateResponse({ ...botResponse, key: responseKey }, (error) => {
+            if (error) {
+                if (error.message.match(/E11000/)) {
+                    setRenameError('Response names must be unique');
+                    return;
+                }
+                setRenameError('There was an error saving your response');
+                return;
+            }
+            setRenameError();
+        });
+    };
+
+    const updateSequence = (oldResponse, content) => {
+        const updatedResponse = oldResponse;
+        const activeIndex = oldResponse.values.findIndex(({ lang }) => lang === language);
+        updatedResponse.values[activeIndex].sequence[0].content = content;
+        return updatedResponse;
+    };
+
+    const handleSequenceChange = (updatedSequence) => {
+        const content = safeDump(updatedSequence);
+        if (isNew) {
+            setNewBotResponse(updateSequence(newBotResponse || botResponse, content));
+            return;
+        }
+        upsertResponse(name || botResponse.key, updatedSequence).then((error, result) => {
+            if (error) {
+                console.log(error);
+            }
+        });
+    };
+
+    const handleModalClose = () => {
+        const validResponse = newBotResponse || botResponse;
+        if (!open) return;
+        if ((!isNew || checkResponseEmpty(validResponse)) && !renameError) {
+            refreshBotResponse(`${language}-${name}`); // refresh the content of the response in the visual story editor
+            closeModal();
+            return;
+        } 
+        if (isNew && !checkResponseEmpty(validResponse)) {
+            if (!responseKey.match(/^utter_/)) {
+                setRenameError('Response names must start with "utter_"')
+                return;
+            }
+            insertResponse(validResponse, (err) => {
+                if (!err) closeModal();
+                if (err) {
+                    console.log(err);
+                    if (err.message.match(/E11000/)) {
+                        setRenameError('Response names must be unique');
+                        return;
+                    }
+                }
+            });
+        }
+    };
```

Comment:
seems like there's some duplication in validations

Reply:
duplication has been removed